Inject writing prompt card into reader stream

eoigal opened the pull request:
This PR intends to inject the writing prompt home card into the Reader following stream.

![Image](https://user-images.githubusercontent.com/5634774/212122435-eca3f336-a1b9-4623-a02b-37763673768b.png)

Ref - https://github.com/Automattic/wp-calypso/issues/72015

Things to work out;

- [x] How do we inject an component into the post stream on Reader?
- [x] How often should the prompt card be rendered in that stream?
- [x] What site will link to when rendering the prompt card? (where will the answer be posted?)

eoigal committed:
Inject writing prompt card into reader stream

eoigal committed:
Look up first blog if no selected site id found

eoigal committed:
hide ellipse menu in reader stream

roo2 commented:
> What site will link to when rendering the prompt card? where will the answer be posted?
I think there are a few options:
* The default site
* The site selector page?
* The "first blogging site" using similar logic to what we used for the notifications

I'm in favor of linking to the site selector page. You can link to `/post` (with no site) and it will take them to this selector. What do you think @davemart-in ?
<img width="1723" alt="Screen Shot 2023-02-06 at 7 15 48 am" src="https://user-images.githubusercontent.com/22446385/216846283-c3347e5e-59ed-4d44-87d5-d5b0d08042e9.png">

eoigal committed:
Use site selector if no site ID is selected

eoigal commented:
I've updated the patch to link to just `/post` when answering a writing prompt and the site ID is not selected - this will mean the site selector will be shown.

I have a concern with how we track stats with no site ID set though. It will appear as null in cases where the site ID is not selected.

I also added some logic in to space out when we show the prompts in the reader stream. I've space it out more than we do the recommended posts as I think we should but let me know if you think differently.

davemart-in commented:
@eoigal I did a walkthrough. Everything looks really great.

I noticed a reproducible JS error in the editor every time I clicked to write a blog post, but I'm not sure it's related to your PR. Here's what I'm seeing:

https://user-images.githubusercontent.com/5634774/217586382-184104b2-10a6-49f6-8386-c428f6310c37.mp4

```
Block validation: Block validation failed for `core/button` ({name: 'core/button', icon: {…}, keywords: Array(1), attributes: {…}, providesContext: {…}, …}).

Content generated by `save` function:

<div class="wp-block-button has-custom-font-size has-medium-font-size" style="text-transform:uppercase"><a class="wp-block-button__link wp-element-button">Get Directions →</a></div>

Content retrieved from post body:

<div class="wp-block-button" style="text-transform:uppercase"><a class="wp-block-button__link wp-element-button has-custom-font-size has-medium-font-size">Get Directions →</a></div>
```

It seems to be talking about a **Get Directions** button? I have no idea where that is coming from.

Note: I see this error when I click the "Post Answer" button on My Home as well, so this doesn't appear to be stemming from your PR.

creativecoder commented on client/my-sites/customer-home/cards/features/blogging-prompt/prompts-navigation.jsx:
```diff
@@ -24,7 +24,9 @@ const PromptsNavigation = ( { prompts } ) => {
 		return prompts !== undefined ? prompts[ promptIndex ] : null;
 	};
 
-	const getNewPostLink = () => addQueryArgs( editorUrl, { answer_prompt: getPrompt()?.id } );
+	// If no site ID set, go through site selector before rendering post editor
+	const getNewPostLink = () =>
+		addQueryArgs( siteId ? editorUrl : '/post', { answer_prompt: getPrompt()?.id } );
 
 	const goToPreviousStep = () => {
 		let nextIndex = promptIndex - 1;
```
We should pull this out of `client/my-sites/customer-home/cards/features/` if it's going to be used in both My Sites and Reader. Maybe somewhere in `client/components`? What do you think?

eoigal replied:
This is a good idea too, I'll create a task to handle this in a follow up PR. Thanks.

eoigal replied:
https://github.com/Automattic/wp-calypso/issues/73158

creativecoder commented on client/my-sites/customer-home/cards/features/blogging-prompt/index.jsx:
```diff
@@ -18,8 +20,24 @@ import './style.scss';
 	const dispatch = useDispatch();
 	const translate = useTranslate();
-	const siteId = useSelector( ( state ) => getSelectedSiteId( state ) );
+	const selectedSiteId = useSelector( ( state ) => getSelectedSiteId( state ) );
+	let sites = useSelector( ( state ) => getSitesItems( state ) );
+	let siteId = selectedSiteId;
+
+	// If no selected site ID, set site ID to first site with write site_intent
+	if ( siteId === null ) {
+		sites = sortBy( sites, [ 'ID' ] );
+		const blogs = Object.values( sites ).filter( ( site ) => {
+			return site.options?.site_intent === 'write';
+		} );
+		if ( blogs.length > 0 ) {
+			siteId = blogs[ 0 ].ID;
+		}
+	}
+
```
What is the selected site when on reader? Is this just the most recent site viewed in the My Sites section?

I thinking falling back to the `getPrimarySiteId` selector if there are no blogs with write intent (which is a fairly recent thing added to onboarding and likely not present for many users) would be a good addition to this block.

eoigal replied:
Good idea, I'll add a fallback to primary blog id if no blog with write intent is found.

In this case we need a site ID to allow us to get the list of prompts.

When rendering the prompts nav though the selected site ID is used in track stats - I'm not sure how to handle this.

roo2 commented on client/my-sites/customer-home/cards/features/blogging-prompt/index.jsx:
```diff
@@ -1,13 +1,15 @@
 import { Card, Button, Gridicon } from '@automattic/components';
 import classnames from 'classnames';
 import { useTranslate } from 'i18n-calypso';
+import { sortBy } from 'lodash';
```
We're trying to avoid bringing in lodash dependencies p4TIVU-9GZ-p2 (@creativecoder just reminded me on my pr 😄) it looks like sortBy can be replaced by the ES6 `sort` which takes a comparison function

eoigal replied:
Removed.

roo2 commented:
I can see a few further design tweaks to match the design shown in https://github.com/Automattic/wp-calypso/issues/72015:

* Vertical space above and below the prompt should be increased to match the internal padding in the other cards
* Prompt should be left aligned with the text, not with the edge of the cards
* Card should have rounded corners to match the images.
* Post Answer button should also have rounded corners.

<img width="706" alt="Screen Shot 2023-02-09 at 9 39 19 am" src="https://user-images.githubusercontent.com/22446385/217677661-536e6e48-a54a-4686-9e8b-3230dba23ccc.png">

roo2 commented on client/my-sites/customer-home/cards/features/blogging-prompt/index.jsx:
```diff
@@ -18,8 +20,24 @@ import './style.scss';
+	let siteId = selectedSiteId;
+
+	// If no selected site ID, set site ID to first site with write site_intent
+	if ( siteId === null ) {
```
Since siteId is only used to retrieve the blogging prompts, I think that we don't need this method to find the first blogging site, the default site would work just as well. 

For the "Post Answer" link, we're just linking to the site selector regardless.

eoigal replied:
Getting rid of lookup on sites with write intent - doesn't serve much purpose now since not setting selected site id.

roo2 commented on client/reader/stream/utils.js:
```diff
@@ -89,3 +89,45 @@ export function getDistanceBetweenRecs( totalSubs ) {
+
+	return Math.min(
+		Math.max(
+			Math.floor( Math.log( totalSubs ) * Math.LOG2E * 5 - 3 ),
```
It seems too complicated, but also quite genius 😄

roo2 replied:
I think it should generally be less frequent and converge to 50 sooner, here's some outputs for `totalSubs` I just tested in the dev tools panel:

2 => 2  (rounded to 10)
5 => 8  (rounded to 10)
10 => 13
20 => 18
50 => 25
100 => 30
200 => 35
500 => 41

eoigal replied:
Updated this to converge to 50 sooner

roo2 commented on client/state/reader/streams/selectors/get-reader-stream-transformed-items.js:
```diff
@@ -26,6 +31,8 @@ export const getTransformedStreamItems = treeSelect(
 			items = injectRecommendations( items, recs, getDistanceBetweenRecs( follows.length ) );
 		}
 
+		items = injectPrompts( items, getDistanceBetweenPrompts( follows.length ) );
```
Nice that this hook already exists for this 👍

eoigal replied:
Agree! Easier to reuse  existing pattern

eoigal committed:
Update styles to copy design mockup

eoigal committed:
add primary site id as fallback if no selected site or site with write intent is found

eoigal committed:
Converge frequency to 50 sooner

eoigal committed:
Remove lookup on sites with write intent - not needed

eoigal committed:
update comment

davemart-in reviewed (approved):
Visually this looks good to me. Nice work!

creativecoder reviewed (dismissed):
I think there's some kind of bug in the injection.

As I scroll down the feed page, the injected card duplicates and starts appearing in other places, like the top of the page.

<img width="869" alt="image" src="https://user-images.githubusercontent.com/1699996/217915869-82d7cef5-1ee4-4cdf-aa50-0774ee067617.png">

creativecoder commented:
Less critical and more opinion, the inset card looks odd to me. Other items injected use the full width, I think we should do the same with the blogging prompts card.

<img width="745" alt="image" src="https://user-images.githubusercontent.com/1699996/217916629-0529066e-2a87-474d-a460-a65a03ce53d9.png">

<img width="820" alt="image" src="https://user-images.githubusercontent.com/1699996/217916638-82b4eb23-0e49-40e0-b4cd-f57313c52aec.png">

ollierozdarz commented:
> the inset card looks odd to me. Other items injected use the full width, I think we should do the same with the blogging prompts card.

@creativecoder @eoigal yeah, now that I see this in context to the entire page, I agree with the above. Let's extend it back out full-width. I beleive it's just removing the `margin-left: 52px;` below.

.reader-stream__blogging-prompt {
    ~margin-left: 52px;~
}

creativecoder commented on client/reader/stream/utils.js:
```diff
@@ -89,3 +89,45 @@ export function getDistanceBetweenRecs( totalSubs ) {
+		if ( index && index % itemsBetweenPrompts === 0 ) {
+			const promptBlock = {
+				isPromptBlock: true,
+				index: index,
```
I'm not sure if there's any practical impact, but the index passed in the `injectRecommendations` function is a separate `recIndex` that's incremented for each recommendation that's injected. Maybe we should do the same here for prompts?

eoigal replied:
This is a good idea. Will leave it for a followup however.

creativecoder replied:
Since it was a 2 line change, I went ahead and did it in https://github.com/Automattic/wp-calypso/pull/72923/commits/4201cf5cb0fb3fb2fa1adcbfdecc5fed38d57053 -- but feel free to revert that if you think it's problematic.

creativecoder reviewed (commented):
It can be in a follow-up, but a nice improvement to this would be showing a different prompt each time the prompt card is injected, rather that the same prompt over and over again.

eoigal committed:
Revert to full-width for card in reader

davemart-in commented:
I tried replicating Gran't bug report. I tried:

- Scrolling slowly
- Scrolling quickly
- Scrolling multiple times down and up the page in the same page load
- Scrolling REALLY far (10+ instances of the writing prompt component)
- Scrolling via the keyboard down arrow

None of which seemed to replicate the bug or produce any sort of JS error.

creativecoder commented:
In case it helps, the account showing the bug is a non-a11n test account, and the default feed has a lot of "Feed: All Latest" posts included in it.

creativecoder commented:
FYI, "Feed: All Latest" is actually the main feed from wired.com.

eoigal committed:
Fix bug with duplicate null keys

creativecoder commented on client/my-sites/customer-home/cards/features/blogging-prompt/index.jsx:
```diff
@@ -18,8 +19,19 @@ import './style.scss';
+	}
+
 	const siteSlug = useSelector( ( state ) => getSelectedSiteSlug( state ) );
+	const notificationSettingsLink = '/me/notifications' + siteSlug ? '#' + siteSlug : '';
```
When I go back to use this card on My Home, the "Manage Notifications" link doesn't work any more.

creativecoder reviewed (commented):
Nice job tracking down the `key` prop issue--I no longer see a console error or duplicated prompt cards with the fix you added.

Just one small problem I'm having with the Notifications link, otherwise this is all now working as expected for me.

creativecoder commented:
For anyone following along, the key prop bug only seemed to be reproducible using a non-a11n account, for unknown reasons. But it looks to be fixed by https://github.com/Automattic/wp-calypso/pull/72923/commits/022f97750a5abcde7625ca31dc2b72953b706b70. I'm glad we caught that and were able to figure out a fix!

creativecoder commented:
@eoigal I created these follow-up issues that came out of reviews/discussions on this PR

- https://github.com/Automattic/wp-calypso/issues/73257
- https://github.com/Automattic/wp-calypso/issues/73258

eoigal committed:
fix notification link url
